File: compilers/python/tests/test_parser_move.py
"""Tests for the .runar.move parser — Python port of the Go reference
``parser_move_test.go``.

Each test pins a representative slice of the Move-style surface so the
Python parser's per-format behaviour is verified explicitly.
"""

from __future__ import annotations

import pytest

from runar_compiler.frontend.parser_dispatch import parse_source


P2PKH_SOURCE = """\
module P2PKH {
    use runar::SmartContract;
    use runar::hash160;
    use runar::checkSig;

    struct P2PKH has SmartContract {
        pub_key_hash: Addr,
    }

    public fun unlock(contract: &P2PKH, sig: Sig, pub_key: PubKey) {
        assert!(hash160(pub_key) == contract.pub_key_hash);
        assert!(checkSig(sig, pub_key));
    }
}
"""


class TestMoveP2PKH:
    def test_parses_to_expected_contract_shape(self):
        result = parse_source(P2PKH_SOURCE, "P2PKH.runar.move")
        assert result.errors == [], result.error_strings()
        c = result.contract
        assert c is not None
        assert c.name == "P2PKH"
        assert c.parent_class == "SmartContract"


class TestMoveProperties:
    def test_properties_and_methods(self):
        source = """\
module Adder {
    use runar::SmartContract;

    struct Adder has SmartContract {
        target: bigint,
    }

    public fun verify(contract: &Adder, a: bigint, b: bigint) {
        assert!(a + b == contract.target);
    }
}
"""
        result = parse_source(source, "Adder.runar.move")
        assert result.errors == [], result.error_strings()
        c = result.contract
        assert c is not None
        assert len(c.properties) >= 1
        assert c.properties[0].name == "target"
        assert len(c.methods) >= 1
        assert c.methods[0].name == "verify"


class TestMoveStateful:
    def test_stateful_counter(self):
        source = """\
module Counter {
    use runar::StatefulSmartContract;

    resource struct Counter {
        count: &mut Int,
    }

    public fun increment(contract: &mut Counter) {
        contract.count = contract.count + 1;
    }
}
"""
        result = parse_source(source, "Counter.runar.move")
        assert result.errors == [], result.error_strings()
        c = result.contract
        assert c is not None
        assert c.parent_class == "StatefulSmartContract"


class TestMoveRejection:
    def test_invalid_syntax_produces_error(self):
        source = """\
module {
    // missing name
}
"""
        result = parse_source(source, "bad.runar.move")
        assert result.contract is None or len(result.errors) > 0


class TestMoveMultipleMethods:
    def test_two_public_functions(self):
        source = """\
module Multi {
    use runar::SmartContract;

    struct Multi has SmartContract {
        x: bigint,
    }

    public fun method1(contract: &Multi, a: bigint) {
        assert!(a == contract.x);
    }

    public fun method2(contract: &Multi, b: bigint) {
        assert!(b == contract.x);
    }
}
"""
        result = parse_source(source, "Multi.runar.move")
        assert result.errors == [], result.error_strings()
        c = result.contract
        assert c is not None
        assert len(c.methods) == 2


# ---------------------------------------------------------------------------
# R-183 (CL-BUG-063): the seven tiers' Move type tables had drifted.
#
# Measured before the fix, across all seven Move frontends:
#
#   Bigint  -> bigint      accepted by 3 of 7 (python, ruby, java)
#   Bytes   -> ByteString  accepted by 2 of 7 (go, rust)
#   address -> Addr        accepted by 4 of 7 (ts, go, rust, ruby)
#
# The same .runar.move source therefore parsed in some tiers and silently
# degraded to a custom type in others. This tier was missing `Bytes` and
# `address`.
# ---------------------------------------------------------------------------
import pytest

from runar_compiler.frontend.parser_move import _move_map_type


@pytest.mark.parametrize(
    "alias,want",
    [
        ("Int", "bigint"), ("Bigint", "bigint"), ("u64", "bigint"),
        ("Bool", "boolean"), ("bool", "boolean"),
        ("vector", "ByteString"), ("Bytes", "ByteString"),
        ("address", "Addr"),
    ],
)
def test_r183_move_type_alias_parity(alias, want):
    mapped = _move_map_type(alias)
    assert getattr(mapped, "name", None) == want, (
        f"{alias} mapped to {mapped!r}; every tier must map it to {want}"
    )

File: compilers/ruby/test/test_parser_move.rb
# frozen_string_literal: true

require_relative 'test_helper'

# Tests for the .runar.move (Move-style) parser. Mirrors the Go suite at
# compilers/go/frontend/parser_move_test.go so all 7 compilers exercise the
# same surface shape on the same inputs.

class TestParserMove < Minitest::Test
  require 'runar_compiler/frontend/parser_move'
  include RunarCompiler::Frontend

  def parse(source, file_name = 'Test.runar.move')
    RunarCompiler.send(:_parse_source, source, file_name)
  end

  # ---------------------------------------------------------------------------
  # Basic Move-style P2PKH
  # ---------------------------------------------------------------------------

  def test_parses_p2pkh_contract
    source = <<~MOVE
      module P2PKH {
          use runar::SmartContract;
          use runar::hash160;
          use runar::checkSig;

          struct P2PKH has SmartContract {
              pub_key_hash: Addr,
          }

          public fun unlock(contract: &P2PKH, sig: Sig, pub_key: PubKey) {
              assert!(hash160(pub_key) == contract.pub_key_hash);
              assert!(checkSig(sig, pub_key));
          }
      }
    MOVE

    result = parse(source, 'P2PKH.runar.move')
    assert_empty result.errors.map(&:format_message),
                 'P2PKH should parse without errors'
    refute_nil result.contract

    c = result.contract
    assert_equal 'P2PKH', c.name
    assert_equal 'SmartContract', c.parent_class
  end

  # ---------------------------------------------------------------------------
  # Properties + methods
  # ---------------------------------------------------------------------------

  def test_properties_and_methods
    source = <<~MOVE
      module Adder {
          use runar::SmartContract;

          struct Adder has SmartContract {
              target: bigint,
          }

          public fun verify(contract: &Adder, a: bigint, b: bigint) {
              assert!(a + b == contract.target);
          }
      }
    MOVE

    result = parse(source, 'Adder.runar.move')
    assert_empty result.errors.map(&:format_message)

    c = result.contract
    refute_nil c
    refute_empty c.properties
    assert_equal 'target', c.properties[0].name
    refute_empty c.methods
    assert_equal 'verify', c.methods[0].name
  end

  # ---------------------------------------------------------------------------
  # Stateful resource struct
  # ---------------------------------------------------------------------------

  def test_stateful_contract
    source = <<~MOVE
      module Counter {
          use runar::StatefulSmartContract;

          resource struct Counter {
              count: &mut Int,
          }

          public fun increment(contract: &mut Counter) {
              contract.count = contract.count + 1;
          }
      }
    MOVE

    result = parse(source, 'Counter.runar.move')
    assert_empty result.errors.map(&:format_message)

    c = result.contract
    refute_nil c
    assert_equal 'StatefulSmartContract', c.parent_class
  end

  # ---------------------------------------------------------------------------
  # Invalid syntax → errors or nil contract
  # ---------------------------------------------------------------------------

  def test_invalid_syntax_errors
    source = <<~MOVE
      module {
          // missing name
      }
    MOVE

    result = parse(source, 'bad.runar.move')
    invalid = result.contract.nil? || !result.errors.empty?
    assert invalid, 'expected errors for invalid Move-style syntax'
  end

  # ---------------------------------------------------------------------------
  # Multiple public functions
  # ---------------------------------------------------------------------------

  def test_multiple_methods
    source = <<~MOVE
      module Multi {
          use runar::SmartContract;

          struct Multi has SmartContract {
              x: bigint,
          }

          public fun method1(contract: &Multi, a: bigint) {
              assert!(a == contract.x);
          }

          public fun method2(contract: &Multi, b: bigint) {
              assert!(b == contract.x);
          }
      }
    MOVE

    result = parse(source, 'Multi.runar.move')
    assert_empty result.errors.map(&:format_message)

    c = result.contract
    refute_nil c
    assert_equal 2, c.methods.length
  end
end
